File: euler9.js
/* euler9.js in JavaScript
 *
 * Running it with node.js from the command line.
 *   $ time node euler9.js
 *   (a,b,c)=200,375,425; a*b*c=31875000
 *   user: 0m0.045s
 *
 * Jim Mahoney | cs.marlboro.edu | April 2014 | MIT License
 */

var pythagorean = function(a,b,c){
    return a*a + b*b == c*c;
}

var search = function(perimeter){
    for (var a = 1; a < 1 + perimeter/3; a++){
	for (var b = a+1; b < 1 + perimeter/2; b++){
	    var c = perimeter - a - b;
	    if (c > 0 && pythagorean(a,b,c)){
		return [a,b,c];
	    }
	}
    }
}

var answer = search(1000);
var product = answer[0] * answer[1] * answer[2]
console.log("(a,b,c)=" + answer + "; a*b*c=" + product)

File: euler9.rb
# euler9.rb
#
#  $ time ruby euler9.rb
#  (a,b,c)=200,375,425; a*b*c = 31875000 
#  user 0m0.724s
#
# Jim Mahoney | cs.marlboro.edu | April 2014 | MIT License

def pythagorean(a,b,c)
  return a*a + b*b == c*c
end

def search(perimeter)
  (1 .. perimeter).each do |a|
    ((1+a) .. perimeter).each do |b|
      c = perimeter - a - b
      if a < b and c > 0 and pythagorean(a,b,c) then
        return [a, b, c]
      end
    end
  end
end

def answer(perimeter)
  abc = search perimeter
  product = abc.inject(:*)
  return "(a,b,c)=#{ abc.join(',') }; a*b*c = #{ product } \n"
end

print( answer 1000 )
